### PR TITLE
Generate trx from scratch

### PR DESCRIPTION
**Pull request by frheault**
Here is a script to build a TRX from a collection of NPY. This script will assemble/cast the NPY into a valid TRX file.
This is still a work in progress, but I have example data [here](https://drive.google.com/file/d/1lEb37KySym0a6_DrxTlWuLRGKz80UbPT/view?usp=sharing).

A reference, positions (CSV), or positions+offsets must be provided. As much metadata as desired can be added.
Due to the support in data type and for simplicity, this script does load arrays in memory.

@anibalsolon When the data is uploaded to BrainLife and the fetcher modified in the code to replace the GDrive.
Could I get a small tutorial on how to add data to BrainLife and fetch it?

I will keep working on the script and improve edge-case and docstring until there is a PR to replace the GDown fetcher by BrainLife. I have a simple test in mind to verify if this script works as intended, but I will wait for issue #25 to be solved.

### CODE REVIEW

**Comment by anibalsolon** [comment]
Hey @frheault 

BL currently provides with a CLI tool as programming interface.To use the CLI, you can follow the instructions here: https://brainlife.io/docs/cli/install/
TL'DR: install Node.js and `npm i -g brainlife`
A Python implementation of CLI and lib to interact with BL may arrive in the next months 🤞🏻

I've set-up in the [org's secrets](https://github.com/organizations/tee-ar-ex/settings/secrets/actions) the BL user and password created for TRX use. So to log in in the CLI tool, you use:

```bash
bl login --username=$BL_USERNAME --password=$BL_PASSWORD
```

This is a small script on how to fetch data:

```bash
# Project ID
PID=6289982fd0697cf1eae8e88d

# subjects: sub-complete, sub-medium, sub-one, sub-resize, sub-small
# datatypes: neuro/track/trx, neuro/track/tck, neuro/track/trk
ID=$(
    bl data query \
        -p $PID \
        -d neuro/track/trx \
        --subject sub-complete \
        --json | jq -r '.[] | ._id'
)

# Download data
bl data download -i $ID -d data

# Move the file to a more meaningful name (and to avoid overwrite)
mv data/track.trx data/complete.trx
```

and now to upload, pretty easy:

```bash
# Project ID
PID=6289982fd0697cf1eae8e88d

bl data upload --datatype neuro/track/tck --project $PID --subject sub-0001 --track track.tck
bl data upload --datatype neuro/track/trk --project $PID --subject sub-0001 --track track.trk
bl data upload --datatype neuro/track/trx --project $PID --subject sub-0001 --track track.trx
```

**Comment by neurolabusc** [comment]
Can I suggest we share TRX data sets using public repositories that do not require a password? Some datasets could be used for regression testing. For example, every commit to [dcm2niix](https://github.com/rordenlab/dcm2niix) uses AppVeyor and Travis to explicitly check conversion of the data in the [dcm_qa](https://github.com/neurolabusc/dcm_qa/tree/2f59a065686f1bdab52f8c4c529e1e2e7e4c133f), dcm_qa_nih, dcm_qa_uih submodules. Each dcm_qa tests different features (e.g. manufacturer). One limitation of Github repositories are file size limits. For larger datasets, I use [OSF](https://osf.io/brvak/wiki/home/) which allows you to include huge datasets and a wiki to describe the rationale.

If there is a concern about sharing raw DTI data due to HIPAA, consider using one of the public DTI datasets. The dcm2niix wiki [provides examples from Siemens, Philips and GE](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage#Diffusion_Tensor_Imaging).

**Comment by skoudoro** [comment]
+1 for @neurolabusc suggestions. I agree that we should avoid using password. 

He might also be over complicated if an user want to test it locally with the provided dataset

**Comment by arokem** [comment]
Yeah. Fair enough. 

@francopestilli : is there a way to create a dataset that don't require authentication on brainlife? Otherwise, I think that we can also use zenodo or figshare for this data.

File: trx/workflows.py [hunk 1]
```diff
@@ -215,3 +222,116 @@ def tractogram_visualize_overlap(in_tractogram, reference, remove_invalid=True):
...
+            positions = load_matrix_in_any_format(positions)
+            offsets = load_matrix_in_any_format(offsets)
+            lengths = _compute_lengths(offsets, len(positions))
+            streamlines = ArraySequence()
+            streamlines._data = positions
+            streamlines._offsets = offsets
+            streamlines._lengths = lengths
+
+        space, origin = get_reverse_enum(space_str, origin_str)
+        sft = StatefulTractogram(streamlines, reference, space, origin)
```
Comment (arokem) [inline trx/workflows.py:251]: Do you want to check if DIPY is available before running this line?

File: scripts/tff_generate_trx_from_scratch.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Generate TRX file from a collection of CSV, TXT or NPY files by individually
+specifiying positions, offsets, data_per_vertex, data_per_streamlines,
+groups and data_per_group. Each file must have it data type specified by the
```
Comment (arokem) [inline scripts/tff_generate_trx_from_scratch.py:7]:
```suggestion
groups and data_per_group. Each file must have its data type specified by the
```

File: scripts/tff_generate_trx_from_scratch.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+users.
+
+A reference file must be provided (NIFTI) and the option --verify_invalid will
+remove invalid streamlines (outside of the bounding box in VOX space).
+
+All dimensions (nbr_vertices and nbr_streamlines) and groups/dpg must match
+otherwise the script will (likely) crash.
+
+Each instance of --dps, --dps, --groups require 2 arguments (FILE, DTYPE).
```
Comment (arokem) [inline scripts/tff_generate_trx_from_scratch.py:16]: It this correct?

```suggestion
Each instance of --dps, --dpv, --groups require 2 arguments (FILE, DTYPE).
```

**Review by arokem — COMMENTED** [review]
Looks like it's heading in the right direction. I spotted a couple of small things.

**Comment by arokem** [comment]
Maybe let's go ahead and upload benchmark data to figshare for now? We can always shift to brainlife if it becomes possible to get the data without authentication.

**Comment by francopestilli** [comment]
hi @arokem see ehre: https://brainlife.io/docs/cli/download/

We require login. This is because the data, well contains brains! :-)
what is the issue with login?

We can create a project, add our users and do one fo the following

See examples:
```
$ bl data download -i 5a050ad5d76a2a002737e572
Download dataset to 5a050ad5d76a2a002737e572
100% done

$ ls 5a050ad5d76a2a002737e572
t1.nii.gz
```

```
bl login

bl pub query -q o3d --json | jq .[0].releases
[
  {
    "removed": false,
    "_id": "5bab993aa918ae0027024192",
    "name": "1",
    "create_date": "2018-09-01T00:00:00.000Z"
  },
  {
    "removed": true,
    "_id": "5bb2437c19fdb40027307f7b",
    "name": "",
    "create_date": "2018-10-01T00:00:00.000Z"
  },
  {
    "removed": false,
    "_id": "5c0ff604391ed50032b634d1",
    "name": "2",
    "create_date": "2018-12-11T00:00:00.000Z"
  }
]
```

**Comment by francopestilli** [comment]
hi @arokem see here: https://brainlife.io/docs/cli/download/
 
We require login. This is because the data, well contains brains! :-) what is the issue with login?
 
We can create a project, add our users and do one fo the following
 
See examples:
 
 ```
$ bl data download -i 5a050ad5d76a2a002737e572
Download dataset to 5a050ad5d76a2a002737e572
100% done
 
$ ls 5a050ad5d76a2a002737e572
t1.nii.gz
```
 
```
bl login
 
bl pub query -q o3d --json | jq .[0].releases
[
  {
    "removed": false,
    "_id": "5bab993aa918ae0027024192",
    "name": "1",
    "create_date": "2018-09-01T00:00:00.000Z"
  },
  {
    "removed": true,
    "_id": "5bb2437c19fdb40027307f7b",
    "name": "",
    "create_date": "2018-10-01T00:00:00.000Z"
  },
  {
    "removed": false,
    "_id": "5c0ff604391ed50032b634d1",
    "name": "2",
    "create_date": "2018-12-11T00:00:00.000Z"
  }
]
```

> Yeah. Fair enough.
> 
> @francopestilli : is there a way to create a dataset that don't require authentication on brainlife? Otherwise, I think that we can also use zenodo or figshare for this data.

The message below
https://github.com/tee-ar-ex/trx-python/pull/32#issuecomment-1138507901

> hi @arokem see ehre: https://brainlife.io/docs/cli/download/
> 
> We require login. This is because the data, well contains brains! :-) what is the issue with login?
> 
> We can create a project, add our users and do one fo the following
> 
> See examples:
> 
> ```
> $ bl data download -i 5a050ad5d76a2a002737e572
> Download dataset to 5a050ad5d76a2a002737e572
> 100% done
> 
> $ ls 5a050ad5d76a2a002737e572
> t1.nii.gz
> ```
> 
> ```
> bl login
> 
> bl pub query -q o3d --json | jq .[0].releases
> [
>   {
>     "removed": false,
>     "_id": "5bab993aa918ae0027024192",
>     "name": "1",
>     "create_date": "2018-09-01T00:00:00.000Z"
>   },
>   {
>     "removed": true,
>     "_id": "5bb2437c19fdb40027307f7b",
>     "name": "",
>     "create_date": "2018-10-01T00:00:00.000Z"
>   },
>   {
>     "removed": false,
>     "_id": "5c0ff604391ed50032b634d1",
>     "name": "2",
>     "create_date": "2018-12-11T00:00:00.000Z"
>   }
> ]
> ```

**Comment by arokem** [comment]
The issue with authentication is that it puts a burden on potential developers and users of the software that have no interest in brainlife per se to get an account on brainlife in order to run the standard tests of the software. For example, getting the software from GitHub does not require a GitHub account. Getting data from figshare requires no figshare account. I understand that for many cases you need to protect the data on brainlife, by requiring an account and authentication, but arguably this is not one of these cases.

**Comment by anibalsolon** [comment]
Figshare is good for me, I can upload it tomorrow. Regarding Brainlife, one solution to be developed could be to allow a project to be marked as public by its owner to allow open access.

**Comment by frheault** [comment]
@anibalsolon Would it be possible to (once on Figshare) to do a PR to replace the fetcher for the tests?

**Comment by anibalsolon** [comment]
For sure!

File: scripts/tff_generate_trx_from_scratch.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    --dpv test_npy/dpv_cx.npy uint8
+    --dpv test_npy/dpv_cy.npy uint8
+    --dpv test_npy/dpv_cz.npy uint8  
+    --dps test_npy/dps_algo.npy uint8 
+    --dps test_npy/dps_cw.npy float64 
+    --groups test_npy/g_AF_L.npy int32 
+    --groups test_npy/g_AF_R.npy int32 
+    --dpg g_AF_L test_npy/dpg_AF_L_mean_fa.npy float32 
+    --dpg g_AF_R test_npy/dpg_AF_R_mean_fa.npy float32 
+    --dpg g_AF_L test_npy/dpg_AF_L_volume.npy float32
```
Comment (arokem) [inline scripts/tff_generate_trx_from_scratch.py:36]: This needs `\` for the line-breaks, right?

```suggestion
tff_generate_trx_from_scratch.py fa.nii.gz generated.trx -f \
    --positions test_npy/positions.npy --positions_dtype float16 \
    --offsets test_npy/offsets.npy --offsets_dtype uint32 \
    --dpv test_npy/dpv_cx.npy uint8 \
    --dpv test_npy/dpv_cy.npy uint8 \
    --dpv test_npy/dpv_cz.npy uint8  \
    --dps test_npy/dps_algo.npy uint8 \ 
    --dps test_npy/dps_cw.npy float64  \
    --groups test_npy/g_AF_L.npy int32 \
    --groups test_npy/g_AF_R.npy int32 \
    --dpg g_AF_L test_npy/dpg_AF_L_mean_fa.npy float32 \
    --dpg g_AF_R test_npy/dpg_AF_R_mean_fa.npy float32 \
    --dpg g_AF_L test_npy/dpg_AF_L_volume.npy float32 
```

File: trx/tests/test_memmap.py [hunk 5]
```diff
@@ -121,24 +126,25 @@ def test__create_memmap(basename, create, expected):
...
-        ("trx/tests/data/small.trx", True, False),
-        ("trx/tests/data/small_fldr.trx", False, False),
-        ("trx/tests/data/dontexist.trx", False, True),
+        ("small_compressed.trx", False, False),
+        ("small.trx", True, False),
+        ("small_fldr.trx", False, False),
+        ("dontexist.trx", False, True),
     ],
 )
 def test__load(path, check_dpg, value_error):
```
Comment (arokem) [inline trx/tests/test_memmap.py:146]: Does this intentionally have two underscores? 
```suggestion
def test_load(path, check_dpg, value_error):
```

**Review by arokem — COMMENTED** [review]
Looks good to me! I think the only remaining item here is the missing docstrings in `utils.py`. Or am I missing something?

**Comment by arokem** [comment]
Just checking in again: is this still WIP? What's the remaining TODOs here? Just the docstrings?

**Comment by frheault** [comment]
Sorry I was moving and now OHBM, I will finish that this week hopefully !
